File: src/skills.mjs
// §10 Inventario de skills. Carpetas con instrucciones reutilizables que el
// harness carga al construir cada llamada. Aqui va lo estable y repetido; lo
// que cambia en cada capitulo viaja en el paquete de contexto (§7), no aqui.
//
// Una skill es texto que lee un modelo, nunca fuente de verdad para el codigo:
// numeros y umbrales viven en config.json (§12).
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { join } from 'node:path';

/** Que skill carga cada rol de §5. */
export const SKILLS_POR_ROL = {
  investigador: ['formato-dossier'],
  arquitecto: ['formato-fichas'],
  escritor: ['formato-paquete-contexto', 'estilo-prosa'],
  validador: ['rubricas-validador'],
  cronista: ['formato-fichas'],
  editor_global: [],
};

function frontmatter(bruto) {
  const m = bruto.match(/^---\n([\s\S]*?)\n---\n?/);
  if (!m) return { meta: {}, cuerpo: bruto };
  const meta = {};
  for (const linea of m[1].split('\n')) {
    const par = linea.match(/^([\w-]+):\s*(.*)$/);
    if (par) meta[par[1]] = par[2].trim();
  }
  return { meta, cuerpo: bruto.slice(m[0].length) };
}

export function cargarSkill(nombre, raiz = 'skills') {
  const ruta = join(raiz, nombre, 'SKILL.md');
  if (!existsSync(ruta)) throw new Error(`skill no encontrada: ${ruta}`);
  const { meta, cuerpo } = frontmatter(readFileSync(ruta, 'utf8'));
  return { nombre, meta, cuerpo: cuerpo.trim(), ruta };
}

export function listarSkills(raiz = 'skills') {
  if (!existsSync(raiz)) return [];
  return readdirSync(raiz, { withFileTypes: true })
    .filter((e) => e.isDirectory() && existsSync(join(raiz, e.name, 'SKILL.md')))
    .map((e) => cargarSkill(e.name, raiz));
}

/**
 * Instrucciones de un rol: su definicion de agente mas las skills que le tocan.
 * El agente vive en agentes/<rol>.md y no lleva numeros: los umbrales que
 * necesite se le inyectan desde config al construir la llamada.
 */
export function instruccionesDe(rol, { raizAgentes = 'agentes', raizSkills = 'skills' } = {}) {
  const rutaAgente = join(raizAgentes, `${rol.replace(/_/g, '-')}.md`);
  if (!existsSync(rutaAgente)) throw new Error(`agente no encontrado: ${rutaAgente}`);
  const { meta, cuerpo } = frontmatter(readFileSync(rutaAgente, 'utf8'));
  const skills = (SKILLS_POR_ROL[rol] ?? []).map((s) => cargarSkill(s, raizSkills));
  const partes = [cuerpo.trim()];
  for (const skill of skills) {
    partes.push(`# Skill: ${skill.nombre}\n\n${skill.cuerpo}`);
  }
  return { rol, meta, skills: skills.map((s) => s.nombre), texto: partes.join('\n\n---\n\n') };
}

File: novela/skills.py
# §10 Inventario de skills. Carpetas con instrucciones reutilizables que el
# harness carga al construir cada llamada. Aqui va lo estable y repetido; lo
# que cambia en cada capitulo viaja en el paquete de contexto (§7), no aqui.
#
# Una skill es texto que lee un modelo, nunca fuente de verdad para el codigo:
# numeros y umbrales viven en config.json (§12).
import re
from pathlib import Path

# Que skill carga cada rol de §5.
SKILLS_POR_ROL = {
    'investigador': ['formato-dossier'],
    'arquitecto': ['formato-fichas'],
    'escritor': ['formato-paquete-contexto', 'estilo-prosa'],
    'validador': ['rubricas-validador'],
    'cronista': ['formato-fichas'],
    'editor_global': [],
}


def _frontmatter(bruto):
    m = re.match(r'^---\n(.*?)\n---\n?', bruto, re.DOTALL)
    if not m:
        return {}, bruto
    meta = {}
    for linea in m.group(1).split('\n'):
        par = re.match(r'^([\w-]+):\s*(.*)$', linea)
        if par:
            meta[par.group(1)] = par.group(2).strip()
    return meta, bruto[m.end():]


def cargar_skill(nombre, raiz='skills'):
    ruta = Path(raiz) / nombre / 'SKILL.md'
    if not ruta.exists():
        raise FileNotFoundError('skill no encontrada: {}'.format(ruta))
    meta, cuerpo = _frontmatter(ruta.read_text(encoding='utf-8'))
    return {'nombre': nombre, 'meta': meta, 'cuerpo': cuerpo.strip(), 'ruta': str(ruta)}


def listar_skills(raiz='skills'):
    base = Path(raiz)
    if not base.exists():
        return []
    return [cargar_skill(d.name, raiz) for d in sorted(base.iterdir())
            if d.is_dir() and (d / 'SKILL.md').exists()]


def instrucciones_de(rol, raiz_agentes='agentes', raiz_skills='skills'):
    """Instrucciones de un rol: su definicion de agente mas las skills que le tocan.

    El agente vive en agentes/<rol>.md y no lleva numeros: los umbrales que
    necesite se le inyectan desde config al construir la llamada.
    """
    ruta_agente = Path(raiz_agentes) / '{}.md'.format(rol.replace('_', '-'))
    if not ruta_agente.exists():
        raise FileNotFoundError('agente no encontrado: {}'.format(ruta_agente))

    meta, cuerpo = _frontmatter(ruta_agente.read_text(encoding='utf-8'))
    skills = [cargar_skill(s, raiz_skills) for s in SKILLS_POR_ROL.get(rol) or []]

    partes = [cuerpo.strip()]
    for skill in skills:
        partes.append('# Skill: {}\n\n{}'.format(skill['nombre'], skill['cuerpo']))

    return {
        'rol': rol,
        'meta': meta,
        'skills': [s['nombre'] for s in skills],
        'texto': '\n\n---\n\n'.join(partes),
    }
